### ISSUE TITLE
Determine and document how request/response metadata is generated in openapi schemas

### ISSUE THREAD

**Issue by mikekistler** [microsoft/cadl#182]
The openapi3 emitter allows certain request/response metadata to be included in models and labeled with decorators, e.g:
- `@path`
- `@query`
- `@header`
- `@statusCode`

Currently any properties with these decorators are not included the schema generated for the model by the openapi3 emitter.

A specific case:
```
model Key {
  @path
  keyId: string;

  @query
  apiVersion: string;

  @statusCode
  statusCode: 200 | 201;

  @header
  contentType: "application/json";

  @header
  Etag: string;

  keyData: string;
}
```
produces the schema:
```
      "Key": {
        "type": "object",
        "properties": {
          "keyData": {
            "type": "string"
          }
        },
        "required": [
          "keyData"
        ]
      }
    }
```

To my knowledge, this behavior (dropping request/response metadata properties) is not documented and it may not even be what we really want or expect.  Anecdotal evidence from @johanste:

![Oh, and I just ran into a @ path decorated property gets omitted from response body models, btw  Contrary to what I wanted](https://user-images.githubusercontent.com/85643503/149825424-306758c6-0114-4a11-af4e-149700641c56.png)

**Comment by mikekistler** [comment]
Need to also consider `@key`.

JS: @key generally means that it is a path segment and a readOnly property of a model.

So maybe @path is a path segment _is not_ a readOnly property of a model. ?

What about @query?  In an HTTP system, it seems like we should exclude these from the schema, but maybe not for a message based protocol.

BT: @path, @query, @header, @statusCode are never emitted to the schema for openapi.

There is some interaction with @visibility and @key that we should work out.

`writeOnce` - in the path and readOnly property in the model

But there are non-key properties that are writeOnce that can't be represented in OpenApi.

**Comment by markcowl** [comment]
-- explicit option

model Foo  {
  bar: string;
}

model FooRequest {
 @query bar: string, 
  @body body: Foo;
}

model FooResponse {
  @statusCode _: 200;
  @body body: Foo;
}

@post op FooOp( ...FooRequest) : FooResponse;

--- implied option ---

model Foo {
  @query @body  bar: string;
  baz: string;
  metadata:  {
   @header @body eTag: string
  }
}

@post op FooOp(...Foo): Foo  | ErrorResponse;

**Comment by nguerrera** [comment]
Part 1 of documenting some background information on status quo here:

I hit something related while writing the TextAnalytics spec in Cadl.

They have a pattern where there are different APIs that use the same input data in different ways. One endpoint takes the data as query parameters while another API takes the same data in the body.

For example:

- https://docs.microsoft.com/en-us/rest/api/cognitiveservices-textanalytics/3.1preview4/entities-recognition-pii/entities-recognition-pii
- https://docs.microsoft.com/en-us/rest/api/cognitiveservices-textanalytics/3.1preview4/analyze/analyze

The first one lets you recognizee pii from a single document synchronously, while the second one lets you batch multiple analyses asynchronously.

## Example

```jsonc
// POST {Endpoint}/text/analytics/v3.1-preview.4/entities/recognition/pii?&domain=phi&stringIndexType=Utf16CodeUnit
{
  "documents": [
    {
      "id": "1",
      "language": "en",
      "text": "My phone number is 555-555-5555"
    }
  ]
}
```

```jsonc
// POST /text/analytics/v3.1-preview.4/analyze
{
  "analysisInput": {
    "documents": [
      {
        "id": "1",
        "language": "en",
        "text": "My phone number is 555-555-5555"
      }
    ],
    "tasks": {
        "entityRecognitionPiiTasks": [
          {
            "parameters": {
              "domain": "phi",
              "stringIndexType": "Utf16CodeUnit"
            }
          }
        ]
      }
    ]
  }
}
```

In order to reuse the same model for these two cases, I added `PlainData<T>` that effectively strips off query/header/path/body decorators from T.

See https://github.com/microsoft/cadl/pull/126

However, I consider that somewhat of a hack and I think we should get rid of `PlainData<T>` and replace it with the solution we implement here in the TextAnalytics spec.

**Comment by nguerrera** [comment]
Part 2.

We currently strip out path/header/body/query/statusCode unilaterally whenever we turn a Cadl model into an OpenAPI schema.

I think it's worth understanding if that is really necessary. I would argue that the main purpose for this stripping is for the cases where we put these
properties elsewhere than the schema. I believe there are only two such cases: headers and status codes for responses.

For example, this:

```cadl
import "@cadl-lang/rest";
import "@cadl-lang/openapi3";

using Cadl.Http;

@route("/example")
@get
op example(): {
  @header foo: string;
  @statusCode code: 200;
  data: string;
};
```

Becomes:

```jsonc
{
"paths": {
  "/example": {
    "get": {
      "operationId": "example",
      "parameters": [],
      "responses": {
        "200": { // status code moved here
          "description": "Ok",
          "headers": {
            "foo": { // header moved here
              "schema": {
                "type": "string"
              }
            }
          },
          "content": {
            "application/json": {
              "schema": {
                "type": "object",
                "properties": {
                  "data": { // other properties here
                    "type": "string"
                  }
                },
                "required": [
                  "data"
                ],
                "x-cadl-name": "(anonymous model)"
              }
            }
          }
        }
      }
    }
  }
}
```

The other metadata properties apply to parameters, not responses. When we're processing the implicit model that represents the set of parameters to an operation, we emit each as an OpenAPI parameter. We never ask for the schema of this model as its own entity. As such, I think it would just work if we didn't strip out body/query/path/statusCode at all: if the property is a parameter it would have its parameter interpretation altered by the decorator accordingly, and if the property was used elsewhere as a non-parameter, the decorator would just be ignored. That would make the textanalytics example just work.

And then even for `@header` and `@statusCode`, currently these are only valid at the top level of a response object. Anywhere else and they will be stripped without being moved elsewhere. In the design meeting, there was debate that maybe you could have something like `{ headers: { @header foo: string }}`, but that doesn't work today, so if that is the justification for not privileging the top-level in this way then I think we should make it work.

So I think that an alternative to more knobs is:

1. Stop removing body/path/query altogether.
2. Only remove header and statusCode when we're in a context where they are consumed and separated from the schema. Currently this context is the top-level of a response object.

That said, it occurs to me now that one complication with conditionally stripping/not stripping is that we can end up with two definitions for a model with the same name, one stripped and one not. We could make this an error and force you to `model is` another name for the different context in the rare case this matters in practice.

**Comment by mikekistler** [comment]
The "two models with the same name" is the problem I've been concerned about. I suppose `model is` is one way to deal with this but another might be to handle intersection better on parameters and return types.

If we handled intersection of this type `<metadata model> & <body model>` better -- as not creating an anonymous model but instead recognizing that it is just a way to split the metadata from the body -- I think that would be a more natural way to express these models.  Other protocols could just ignore the metadata model.  And OpenAPI emitters could make the body (request or response) a ref and not an equivalent anonymous model.

What I'm proposing would be equivalent to a single model with decorators on all the metadata properties _and_ the body model on a property annotated with @body, but it avoids the `@body` decorator.

**Comment by nguerrera** [comment]
I'd be worried about assigning special meaning to `A & B` distinct from `{...A, ...B}`. Also, how do we know that A is the metadata model and B is the body? Wouldn't it have to be something on A that potentially prevents it from being reused as data elsewhere as in the textanalytics example? I also dislike solutions where A & B is not equivalent to B & A.

Also, @bterlson always notes that whenever we try to firm up this segregation, we end up limiting cross-protocol scenarios where something is metadata for one protocol but plain data for another. I share this concern. That said, if we wanted to go down that route with that downside, I would prefer something based on an unambiguous shape like `{ @metadata: A @body: B }` (strawman) and then to approach terseness  of `&`, this could be templated as `Response<Metadata, Body>`.

**Comment by nguerrera** [comment]
I wonder if projections could help with the cross-protocol problem. Like, again strawman project `Response<Metadata, Body>` to { ..RemoveStatusCode<Metadata>, ...Body } for a hypothetical case where you send http headers as plain fields. So each protocol would have a #toProtocol projection to do stuff like this?

**Comment by nguerrera** [comment]
> Other protocols could just ignore the metadata model.

I mistakenly missed this in my reply above. This could deal with the cross-protocol problem, but I strongly favor different emitters interpreting different shapes and decorators differently vs. reinterpreting primitive language construct like `&`.

**Comment by johanste** [comment]
I just want to call out that, for a HTTP based protocol (ignoring cross protocol issues for now), not all information that goes in the body is necessarily message payload (there can be request/response metadata in there), nor are things that go outside the body necessarily metadata. For example, for blob properties operations, the "message" goes in headers. The mapping needed is very similar to what we have to do in client libraries when we determine what "T" really is in "Response Of T" for our client libraries. 

In other words, logical message != (http) body. So there would presumably have to be some form of protocol specific mapping of properties in the logical message into where they are put in a protocol specific request/response.

**Comment by markcowl** [comment]
Much of the metadata, even what we consider 'payload' metadata is operation-specific and request/response specific

  - ARM the 'name' property gives the resource name, which is used as the last path parameter for resource operations *and* is duplicated in the body, but only in the response body, it is *not* in the body of a PUT or PATCH request.

One mechanism for mapping model properties into request and response payloads is visibility.  'read' = response payload, 'create' = request payload of creation methods, 'update' = request payload of PATCH methods.

- Metadata that is never included in payloads
  - statusCode

- Metadata that is *always* included in payload
  - not sure anything falls in this category
  
- Metadata that is sometimes included in payload
  - query
  - headers
  - path

Unless we can determine hard and fast rules for including metadata in the payload, we need sufficient degress of freedom to allow the author to specify, and we likely want defaults that make sense, like
- path: by default,  in response payload
- header, query: by default not in payload

**Comment by markcowl** [comment]
Additional issues:
- modeling BATCH requests and responses
  - status code in body may be a list of operation status codes
- data vs. metadata
  - metadata = http protocol data outside the body  OR data that is not logically part of the resource
  - operation-specific metadata, like top/skip in list APIs
    - adding metadata should *not* be a breaking change
      - optional headers
      - query parameters
      - need to have a standard, prescribed order to avoid breaking change issues in clients
      - spreading models into parameters might cause an issue
- examples of 'metadata' that are part of a model
  - eTag headers
  - query and header parameters in some data planes

model Foo {
  @header @body eTag: string;
  bar: string;
}

 - 'eTag' included in response?
 - 'eTag' included in logical model?
 - eTag is normally returned in a header, but in some cases, may be in response body
 - verify in how this is represented

model Foo {
  eTag: string;
  bar: string;
}

model FooMetadata {
  @header eTag: string
}

op getFoo(...) : FooMetadata & Foo;  // {...FooMetadata, @body: body: Foo} // model FooResponse {...FooMetadata, @body body: Foo}}

model Foo {
  @query bar?: string;
  baz: string;
}

- @header, query, statusCode in body model means exclude from schema
- @path may require some additional mechanism

**Comment by nguerrera** [comment]
Stepping back, I realize I didn't post any equivalent Cadl for the TextAnalytics example, so let me fix that.
```
model PiiParameters {
   @query stringIndexType: StringIndexType;
   @query domain: "phi" | "none"
 }

model Tasks {
   entityRecognitionPiiTasks: PiiTask[];
}

model PiiTask {,
   parameters: PiiParameters;
}
      
 model Batch {
    anaylysisInput: Document[]
    tasks: Tasks;
}
 
// here we emit the members of PiiParameters as query parameters, we never ask for the schema of PiiParameters as a standalone entity
 op recongizePii(...PiiParameters, @body body: Document): Response;

// here we do see PiiParameters as something we need a schema for, but why must we strip the query parameters in this context? I remain surprised by this behavior.
 op analyze(@body body: Batch): Response;
 ```
 
Can someone provide an actual scenario/example where a `@query` is on a non-parameter property in an operation signature and needs to be removed from an OpenAPI schema? I would argue we need that before we add knobs to control this. It's also not clear to me that knobs would even allow the TextAnalytics sample to work. How would I set them differently in the different contexts? I think we'd end up back at something like `PlainData` to transform a type into one with different knobs. 

So I still think `@path` and `@query` can depend on parameter-or-not context. And even if that is configurable, I think it is still the best default.

I oversimplified earlier including `@body` in this bucket because that can go on parameters and on a response property, but still, we have enough context there: if it's neither being consumed as a body parameter or as the response body, then take it to be a regular data property.

In general, I find it surprising whenever a property decorator causes a property to just disappear from a schema while having zero other impact on the emit. I lso we have enough context there, if it's neither being consumed as a body parameter or as the response body, then take it to be a regular data property.

In general, I find it surprising whenever a property decorator causes a property to just disappear from a schema while having zero other impact on the emit. I haven't understood yet if this is ever required, and it's very surprising behavior to me, but I think I might understand the pushback better if I saw some examples.

**Comment by markcowl** [comment]
- headers make sense, but maybe not query
- parameter ordering: 
op @put createFoo(...KeyParameters<Foo>, Request<Foo>) : Response<Foo> | ErrorResponse; 

model Foo {
  @header  eTag: string;
  bar: string;
}

- consensus on constraints
  - proposed example: model properties should not be removed
  - eTag example
  - isomorphism between parameter lists, response models

**Comment by markcowl** [comment]
The v1 portion of this is handled in https://github.com/microsoft/cadl/issues/267

**Comment by mikekistler** [comment]
Here's an interesting case I just discovered: an operation that returns an "etag" as both a header and as a top-level property.  The cognitive services control plane `createAccount` operation does this.

Does our current proposal support this case?

**Comment by bterlson** [comment]
I don't think it does. Presumably we could add a decorator that you can add to metadata fields that says to keep the property in the schema as well. Would we need to also be able to say which "wins" when unpacking the response into the type we return to the user, or is it enough to always pick one or the other?

**Comment by mikekistler** [comment]
I have an alternate proposal that I think has some merits:

https://gist.github.com/mikekistler/fa2961db03e8249bd103b10442500e85

Feel free to leave comments there or in this issue.

**Comment by nguerrera** [comment]
#506 is a first step on this. 

Interestingly, it actually contributes to both the main proposal and the alternate proposal so is still on common ground. Both proposals actually require that you get good name `Data` for  `<Metadata> & Data`, which #506 addresses already and the same mechanism it uses for that could be used to build out the rest of the main proposal, which is what I'm working on now. We will need to bring this back to design review once we have a working end-to-end for the main proposal.

**Comment by nguerrera** [comment]
Recap since last update:

- A draft implementation of the proposal was done in https://github.com/microsoft/cadl/pull/687
- A write-up of the challenges and open issues with it is here and was discussed in design meeting: https://gist.github.com/nguerrera/8d1843069c09062c1704bb49580afa9a
- In that discussion and subsequent discussions it became clear that we need an approach that does not require so much work in the emitter and that can be reused for other transformations.
- We brainstormed a "realm" feature that would be extracted from projections and could make this work.
- We determined that some simplifications are needed to the proposal to make certain hard problems like "deep deduping" go away.

I am going to close this design issue as having tracked the original proposal and open new issues to break down getting to a landing place on this into steps:

- [ ] Amended metadata design proposal with simplifications
- [ ] "Realm" design proposal
- [ ] Implement simplified metadata proposal
- [ ] Rework metadata implementation to use "realms"
